modules/fluid_properties/src/userobjects/CO2FluidProperties.C

```diff
@@ -0,0 +1,719 @@
+}
+
+Real
+CO2FluidProperties::mu(Real density, Real temperature) const
+{
+  // Check that the input parameters are within the region of validity
+  if (temperature < 216.0 || temperature > 1000.0 || density > 1400.0)
+    mooseError("Parameters out of range in CO2FLuidProperties::mu");
+
+  Real Tstar = temperature / 251.196;
+  const std::vector<Real> a {0.235156, -0.491266, 5.211155e-2, 5.347906e-2, -1.537102e-2};
+  const std::vector<Real> d {0.4071119e-2, 0.7198037e-4, 0.2411697e-16, 0.2971072e-22, -0.1627888e-22};
+
+  // Viscosity in the zero-density limit
+  Real sum = 0.0;
+
+  for (unsigned int i = 0; i < a.size(); ++i)
+    sum += a[i] * std::pow(std::log(Tstar), i);
+
+  Real mu0 = 1.00697 * std::sqrt(temperature) / std::exp(sum);
+
+  // Excess viscosity due to finite density
+  Real mue = d[0] * density + d[1] * std::pow(density, 2) + d[2] * std::pow(density, 6) /
+    std::pow(Tstar, 3) + d[3] * std::pow(density, 8) + d[4] * std::pow(density, 8) / Tstar;
+
+  return (mu0 + mue) * 1e-6; // convert to Pa.s
+}
+
+void
+CO2FluidProperties::mu_drhoT(Real density, Real temperature, Real & mu, Real & dmu_drho, Real & dmu_dT) const
+{
+  // Check that the input parameters are within the region of validity
+  if (temperature < 216.0 || temperature > 1000.0 || density > 1400.0)
+    mooseError("Parameters out of range in CO2FLuidProperties::mu_drhoT");
+
+  Real Tstar = temperature / 251.196;
+  const std::vector<Real> a {0.235156, -0.491266, 5.211155e-2, 5.347906e-2, -1.537102e-2};
+  const std::vector<Real> d {0.4071119e-2, 0.7198037e-4, 0.2411697e-16, 0.2971072e-22, -0.1627888e-22};
+
+  // Viscosity in the zero-density limit. Note this is only a function of T
+  Real sum0 = 0.0;
+
+  for (unsigned int i = 0; i < a.size(); ++i)
+    sum0 += a[i] * std::pow(std::log(Tstar), i);
+
+  Real mu0 = 1.00697 * std::sqrt(temperature) / std::exp(sum0);
+
+  // Excess viscosity due to finite density
+  Real mue = d[0] * density + d[1] * std::pow(density, 2) + d[2] * std::pow(density, 6) /
+    std::pow(Tstar, 3) + d[3] * std::pow(density, 8) + d[4] * std::pow(density, 8) / Tstar;
+
+  Real dmuedrho =  d[0] + 2.0 * d[1] * density + 6.0 * d[2] * std::pow(density, 5) /
+    std::pow(Tstar, 3) + 8.0 * d[3] * std::pow(density, 7) + 8.0 * d[4] * std::pow(density, 7) / Tstar;
+
+  // Use finite difference for derivative wrt T for now, as drho_dT is required
+  // to calculate analytical derivative
+  Real eps = 1.0e-8;
+  Real Teps = temperature * eps;
+  Real mu2T = this->mu(density, temperature + Teps);
+
+  // Viscosity in Pa.s is
+  mu = (mu0 + mue) * 1e-6;
+  dmu_drho = dmuedrho * 1e-6;
+  dmu_dT = (mu2T - mu) / Teps;
+}
+
+Real
+CO2FluidProperties::partialDensity(Real temperature) const
+{
+  // This correlation uses temperature in C
+  Real Tc = temperature - _T_c2k;
+  // The parial volume
+  Real V = 37.51 - 9.585e-2 * Tc + 8.74e-4 * Tc * Tc - 5.044e-7 * Tc * Tc * Tc;
+
+  return 1.e6 * _Mco2 / V;
+}
+
+std::vector<Real>
```

Comment:
`const std::vector<Real> &` will be better unless you plan on changing the constants...